Moving instat_calculation code from R-Instat into a separate R package

lilyclements opened the pull request:
I was too anxious to wait, so I tried it out for the `instatCalculations.R` file, and it worked for the testing I did. 

**The method:**
* Removed `instat_calculation` and `calculation` R6 files
* Removed all instances of `instat_calculation$set`, and `calculation$set`
* Replaced all instances of `instat_calculation$new` with `instatCalculations::instat_calculation$new`

**To-do:**

1. Where do we add in the `devtools::install_github("IDEMSInternational/instatCalculations")` line so that we can be sure that it is correctly loaded into R-Instat (presumably in `Rsetup.R`).
2. Where do we add in the `library(instatCalculations)` R code (presumably in `Rsetup.R`)
3. Currently I have set it up that it runs as the function is `instatCalculations::instat_calculation$new` in our dlg R code. We want to separate this out to be 

```
XXX.SetPackageName(instatCalculations)
XXX.SetRCommand(instat_calculation$new)
```

This is just a draft/test illustration of the approach to do this, and a test to see if it works!

lilyclements commented:
@rdstern if you wanted to view or test, then this is the PR

lilyclements commented on instat/clsQualityControl.vb:
```diff
@@ -66,7 +66,7 @@ Public Class clsQCJumpRCode
 
         strCalcName = strlargestJump
-        clsJumpCalcFunction.SetRCommand("instat_calculation$new")
+        clsJumpCalcFunction.SetRCommand("instatCalculations::instat_calculation$new")
```
E.g., here we want to separate into 

```
clsJumpCalcFunction.SetPackageName("instatCalculations")
clsJumpCalcFunction.SetRCommand("instat_calculation$new")
```

I just did it this way as a quick check to see if it would actually work. I'm very surprised it worked for the testing I did.

rdstern commented:
@lilyclements very exciting.  

I tried, but can't open a dataset?

![image](https://github.com/user-attachments/assets/cf0e3ee3-e918-436e-b97c-c852b139b4ae)

I have a problem on my test version of not having all the packages we need - and am gradually adding them.   But I think this may be different.

rdstern commented:
@lilyclements I tried again, first adding all the packages into my test version.
I still get the same error, so I think it is at your end.  And excited that you said you thought you might know where it is.
I assume to test, I should use the calculator and maybe also the summarise dialog etc?

lilyclements commented:
@rdstern great - can you try again now? I cannot replicate it my end, but I could find the line where that error was occurring (and other places where it would occur in the future), and it should hopefully be fixed now :)

rdstern reviewed (commented):
@lilyclements very exciting.  I report on my tests so far:
a) I had lots of trouble installing instatCalculations in the test version.  I tried many options and nothing seemed to work, including the Import saying the directory didn't exist.  However, I suspect this is a glitch on my version only, so report just in case others have difficulty.  However, I stopped and started again (maybe should have done that earlier) and it now said it was installed and up-to-date.  Whew.  So I started testing.
b) Now I could import from the library - examples of datasets came in fine.  They hadn't before and seemed to need the package, because they complained when it wasn't installed.  Incidenetally I suggest that it and databook be included in the run, so the package name is not needed when used.
c) I tried importing a dataset from a file, and also a new data frame.  All worked fine.  Those are the top 3 dialogs in the File menu.
d) I tried a few calculations - R-Instat calculator and that was ok.
e) I tried the Prepare > Column: Numeric > rounding data, but on a selection of variables from diamonds - so using select as well.  All ok.
f) I then tried the Prepare > Reshape > Column Summarise and got my first error.  I tried simple summaries on our usual small survey data:

![image](https://github.com/user-attachments/assets/2624a524-1582-4ca7-9c70-bd921b4f645f)

g) I assumed that if summarise doesn't work, then the Describe > One Variable > Summarise > Customised would fail, but it is fine!

h) So I went back to the Prepare > Reshape > Summarise dialog.  I found it works fine with the new features, so when there are no factors.  And that was added quite recently.  I wonder if that added something into the code that is not in the version of instatCalculations that you have now?
 
All very promising

Patowhiz commented:
@lilyclements good to see that the calculation code is being separated out. I have many plans that relates to iterating from the original code. Are you looking towards releasing it with documentation or that's a low priority at the moment?

rdstern commented:
@Patowhiz glad you like it.  If we mean the same on documentation, then that's done already - and for all the R-Instat packages, and also available and linked from the help.  So, once working the documentation will also - like the other packages - be available in your maximised window, so without needing the internet links.

In addition, these packages are not yet in CRAN, but are in IDEMS github site, and the import packages dialog has been amended so they can easily be installed.

This is another stage in this huge step forward!

lilyclements commented:
@rdstern you're absolutely right that I hadn't updated `instatCalculations` since we added in `param_list`. It's now updated, so if you update your `instatCalculations` file, this should now work!

(I did update `databook` relatively recently, but, I imagine that will need re-updating since that is such a large package!)

lilyclements commented:
@rdstern I have updated `databook` now (all the `undo` functions and what not were not in there). I will need to  update them again in the future from R-Instat, but, they're up to date as of the Master version today (I've noted down more specifics in the issues on their repos so I can be more precise on where to update from -- e.g., the most recent commit, etc)

rdstern commented:
@lilyclements you said it would be easy to resolve conflicts?

lilyclements commented:
@rdstern thanks for this, yes I've fixed it now

rdstern commented:
@lilyclements so I updated instatCalculations and still get the same unused argument error as shown above?

rdstern commented:
@lilyclements I think you have located the problem:
a) In the test version - using your branch - I updated the instatCalculations package.  It seems to update fine and here is the code in the output Window:

![image](https://github.com/user-attachments/assets/490c36f9-cf55-4d5a-879c-e0d4e8572958)

However, when I return to the importing dialog, the Check remains Green!

I have tried different options, like the upgrade = never, but it always remains green!

I then tried with Version 8.1 and it goes green, as above, after installation!

Aha, I still don't kn ow why it won't work in R-Instat for me, but I tried via RStudio, and now it is brown in R-Instat.  Phew.  Let's see now!

Now I restarted your branch, it is still brown and no error!!  Yippee.  

@N-thony really good if you could check now anyway - even if we don't merge yet.  No danger, because I think you can't merge a branch that is labelled draft.

lilyclements commented:
@rdstern very interesting! This is something we should investigate a bit further for installing packages in the future. I'm not sure where the R-Instat packages store to when installed, and if this is the same or a different location to that of when we do it in RStudio.  Glad that you could test it at least and it's working for you now!

rdstern commented:
@lilyclements Version 0.8.2 is now released.  So we are ready for you to make this big change.  I suggest you change the title of this pull request and then include the remaining packages first?  I assume the functions within the current code will now also be deleted at the same time.  Can we (I mean you?) do this and merge this week?

Once merged, then I assume - if all goes well - all the other open pull requests need to update, so their branch uses the latest merged version.  Then they should be ok again?

rdstern reviewed (commented):
I approved, because it seems to work well.  I was originally thinking we could merge a package at a time.  But I now feel that the merging could induce conflicts in the other open pull requests.  @N-thony is this correct?  If so it will be better to add the other packages and then do a once and forever merge.  With this assumption I'm withdrawing the approval for just the one package.

rdstern reviewed (changes requested):
Include the reamining packages

lilyclements commented:
Not ready yet. Left for me to do:
1. Check labels (see constants.R in databook)
2. Add package name for our functions in [this file](https://github.com/IDEMSInternational/databook/blob/main/R/summary_functions.R)
3. Check for adding `databook` names into the `instatCalculations.R` file
4. Where do we add in the 
```
devtools::install_github("IDEMSInternational/databook.R")
```
so that we can be sure that it is correctly loaded into R-Instat (presumably in Rsetup.R).
5. Where do we add in the `library(databook)` R code (presumably in Rsetup.R).
6. We don't need to run `data_book <- DataBook$new()` on opening because this automatically runs in the `databook` R package now.

rdstern commented:
@lilyclements tell me when it is ready for me to look again.  I tried just now, and it can't find data_book.  I installed that, but it was still not happy.  I assume I am a bit premature?

lilyclements commented:
Hopefully soon - just running some checks myself and fixing bits on my end.

lilyclements commented:
Only premature by a couple of minutes - I think it can now be tested. Do update your `databook` again since I just merged in a new change into it.

rdstern commented:
I report as I go along.
a) Installing databook gave me the same trouble as before.  It appeared to install and then was still green.  I eventually did as above, namely copied the one-line installation script into RStudio and installed from there.  It still didn't work until I closed and re-opened R-Instat.  
b) I assumed I would also have to install instatCalculations, but it said that was already up-to-date.
c) Then I could read in a data frame.  Great.
d) Then went to Prepare > Calculator and it said I needed instatExtras.  So installed that and it seems to install fine (goes brown).
e) Wowsers - it seems to work.  I tried to trick you by using the climatic menu, because I expected that package still needs to be added.  Is it all in Instat extras?

So, it now works for me - seemingly.  What would you like me to try next?

lilyclements commented:
I am quite surprised as to how easily it was to all work. I haven't added in the `instatExtras` and `instatClimatic` bits (which are just stand alone functions), but I can do that now so we call from those packages. That's definitely the much easier part - but a bit more mundane. The perfect type of job for 530pm on a Monday while I'm having a snack (hey, that's now!)


To the tricots point - the `source("...")` bit is just running the functions [here](https://raw.githubusercontent.com/AgrDataSci/ClimMob-analysis/master/modules/01_functions.R). You can try running that script in R-Instat to see where any issues are arising? Otherwise, the R code for the network graph is:

```
network = function(object, ...) {
  
  if (class(object) == "grouped_rankings") {
    object = as.rankings(object)
  }
  
  R = object 
  
  adj = PlackettLuce::adjacency(R)
  
  adj = as.vector(adj)
  
  adj = t(matrix(adj, nrow = ncol(R), ncol = ncol(R)))
  
  dimnames(adj) = list(dimnames(R)[[2]], dimnames(R)[[2]])
  
  adj = btdata(adj, return_graph = TRUE)
  
  netw = adj$graph
  
}
```

lilyclements commented:
@rdstern `instatExtras` is now in there and is hopefully working well. This means it is also now being called into our `databook` package - before we had missed a few bits on where to call it. So, I suggest you update `databook` again.

I'll add in `instatClimatic` shortly (tomorrow?)

rdstern commented:
@lilyclements I couldn't update databook, either in R-Instat or in RStudio:

In RStudio it says:
`Error: Failed to install 'databook' from GitHub:   Missing commas separating Remotes: 'IDEMSInternational/instatExtras     IDEMSInternational/instatClimatic'`
-

rdstern commented:
Well, that's changed things.  When I try to install databook now it says:

![image](https://github.com/user-attachments/assets/71af9eef-37f0-4480-8ac0-673bc142b011)

Hope that's useful too!

So I tried to install climdex.pcic and it said:

![image](https://github.com/user-attachments/assets/0aa7fd10-0c3a-420b-af0a-89dede953394)

And same message when I try to install instatClimatic.

lilyclements commented:
@rdstern I plan to fix the issues in instatClimatic today. This comes from the climdex.pcic package no longer being on CRAN. I’ll let you know once this is sorted.

lilyclements commented:
@rdstern hold off on that for a bit! Doing that has created a new issue for me - I think from updating a package.

lilyclements commented:
@N-thony I am having a new issue here. It works fine in R, but in R-Instat my session is unable to read in data:

![image](https://github.com/user-attachments/assets/f4f3385d-4bb0-429b-b035-6a17a72a12c5)

It is working in R, and I do not understand the R-Instat VB code associated to this - (clsDataBook, sub on line 133). It seems the data is not "refreshing successfully", but I do not understand what that means or how to fix that.

lilyclements commented:
OK, @rdstern, now can you try running the following script:

```
remove.packages(c("instatExtras", "instatClimatic", "databook", "instatCalculations"))

devtools::install_github("IDEMSInternational/instatClimatic", upgrade = "always")
devtools::install_github("IDEMSInternational/instatExtras", upgrade = "always")
devtools::install_github("IDEMSInternational/databook", upgrade = "always")
devtools::install_github("IDEMSInternational/instatCalculations", upgrade = "always")
```

This just removes our four packages, and then reinstalls them back in. We can see if that works for fixing this. If not, then let me know and I can remove instances of `instatClimatic` from `databook` for now.

Note that you cannot do this on this branch - this is because this branch uses the `instatExtras` for the R script window. Can you run this on a different branch?

rdstern commented:
@lilyclements I tried. It doesn't complain on the remove.  Then it installs Extras and Calculations fine.  But the other two give the same error below:

![image](https://github.com/user-attachments/assets/693a0196-209e-44e6-8049-6d23429bee23)

rdstern commented:
@lilyclements I tried the 2 offending lines in RStudio and it went a long way and thensaid I needed build tools, and did I want to install them?  That seems to relate a bit to the error message above?

lilyclements commented on instat/static/InstatObject/R/Rsetup.R:
```diff
@@ -225,8 +225,8 @@ packages_not_loaded <- function() {
 setwd(dirname(parent.frame(2)$ofile))
-source("instat_object_R6.R")
-source("data_object_R6.R")
+#source("instat_object_R6.R")
+#source("data_object_R6.R")
```
Great, I'll "undo" those edits, but later. For now, it's useful for testing!

lilyclements commented:
@rdstern I think in terms of how we want these packages set up that we do not want `databook` to depend on `instatClimatic` anyway. So I have removed that dependency now, which means that `databook` should download without any issues. 

There are inevitably still functions in `databook` that depend on `instatClimatic`, and I will get to these tomorrow or this evening on the train home. This means you may encounter a bug or two until I can remove the functions across, but at the very least it can now download!
(Basically, you may get an error saying it cannot find a function. That's because it was in the `instatClimatic` package but is not anymore. But it will soon be in the `instatExtras` package once I've migrated everything over. Shouldn't take long, but I will need an hour or so to do it!)

rdstern commented:
@lilyclements a report so far. 

a)  install R package dialog.  The databook appeared to work, but still green next time.
b) Tried the command in the script and seemed to work, maybe better.  But reasonably quick.  Then checked with the dialog and it froze.
c) Tried installing databook in RStudio.  Did lots and took quite a time.  But then appeared to work.  At least it didn't ask me to install anything more.
d) So, got bold and tried instatClimatic in RStudio.  That's the same as before.
e) Closed R-Instat and opened again.  Checked on databook, and yippee!

So now starting testing with your databook included.  How exciting is that!!!

Humpf I thought I should check the others and it says I need to install instatExtras.  
So I did, and that went brown nicely.

Now tried a variety of different things.  All seems to be working.

The only "failure" was in the summarise without dropping unused levels - and it did.  I need to check if that is just here, or might already be going wrong in Version 0.8.2!

Just to confirm we have scripts for a lot of our practicals.  Once you have all the packages we could run through those as a good test - couldn't we?  Beryl can help too.

Looking forward and hope you are on the "home run" now?

lilyclements commented:
@rdstern, can you try installing `instatClimatic` now? Let me know if it works and I can send up the visual-studio-related changes

rdstern commented:
Sould we get another machine to try.  We could ask Beryl and maybe others?
We also probably want @ChrisMarsh82 to test, so should we ask him?

lilyclements commented:
@rdstern sounds good. Can you see if there's an error in the output window on Visual Studio when you run it? I don't know what the issue is

![image](https://github.com/user-attachments/assets/1a96cff1-378b-4cca-a75a-938a46afbadc)

In the bottom right hand panel is "Output". Can you send me what that says after you try to install `instatClimatic`?

rdstern commented:
You are right.  It says:

```
Using GitHub PAT from the git credential store.
The thread 0x34c has exited with code 0 (0x0).
Downloading GitHub repo IDEMSInternational/instatClimatic@HEAD
Downloading GitHub repo IDEMSInternational/databook@HEAD
Installing 1 packages: circular
Warning: package 'circular' is in use and will not be installed
Warning: package 'databook' is in use and will not be installed
Installing 1 packages: texmex
Warning: package 'texmex' is in use and will not be installed
Downloading GitHub repo IDEMSInternational/databook@HEAD
Skipping databook, it is already being installed
Installing package into 'C:/Users/rdste/AppData/Roaming/R-Instat/1.0.0/library'
(as 'lib' is unspecified)
ERROR: dependencies 'databook', 'texmex' are not available for package 'instatClimatic'
* removing 'C:/Users/rdste/AppData/Roaming/R-Instat/1.0.0/library/instatClimatic'
R viewer not detected. File saved in location C:\Users\rdste\AppData\Local\Temp\RtmpAFsORx\viewtext8c74294823.txt
```

So I tried:

```
detach("package:texmex", unload=TRUE) 
detach ("package:circular", unload=TRUE)
```
Then it said texmex and circular were not installed, so I installed them!  Then I tried again.  It gave a similar error, but just mentioning databook this time.

So I ran 

`detach ("package:databook", unload=TRUE)  `

Then I tried installing instatClimatic again - and it is nice and brown!!!

So I think I now have them all!

I have started testing!

a) I have opened our usual survey.  That works.
b) I tried a new data frame and that works
c) I tried the column summarise and I now get the 12 summaries when I don't drop unused levels.  So that works again.
d) Other tasks with the survey.  All working so far.

e) But then I tried to load another datasets and got this error:

![image](https://github.com/user-attachments/assets/438bad89-c615-4a27-b10e-ef80c5933d18)

f) Importing an Excel file also doesn't work now:

![image](https://github.com/user-attachments/assets/e0432aaf-6f7b-4e20-8c04-e79a80c54e1c)

rdstern commented:
@lilyclements from the messages above I have been pondering why I had such a problem adding InstatClimatic, and you didn't.  I have found that my test version of R-Instat does not have all the packages pre-loaded.  In particular I assume I may not have had `texmex `or `circular` in my version.  In my testing generally I have been looking for the error message `don't have package xxx ` as a spur to adding it.  If that is the problem it will be ok when using R-Instat normally, because all packages are pre-loaded.

rdstern commented:
@lilyclements all very exciting and I'm just about to try your new versions.  In preparation I have written a little about detach, etc, into #9314 .  In the future I suggest we may want to allow updates of the R-Instat packages, between full updates of R-Instat.  These packages will usually be attached, to we may need to enhance the updating dialog to make that a reasonably easy job.

Now on to testing the new versions of your packages!

a) getting to "all brown"!  I found that instatCalculations and instatClimatic were ok - still up-to-date.  instatExtras and databook needed updating.
b) I first did that in R-Instat.  I had to use detach to get to them "all brown".  But then I still got the same 2 errors above, i.e. the `get_data_book_output_object_names` not found, and the inability to import Excel files.  I also tried with a csv file and got the same medssage as above.
c) So I quit the branch and tried again.  Found databook and instatExtras has gone green again!  So I tried adding them from RStudio - without adding any more packages.  Then, on returning to R-Instat and closing, and starting again, all was brown.  But the same two errors persist.

What am I doing wrong - or not doing?

lilyclements commented:
@rdstern if you run `databook:::get_data_book_output_object_names` in the script window, does it throw an error or give some code? Just trying to understand the situation.
Are you using the most recent PR too - perhaps there's something in there too.

rdstern commented:
@lilyclements running that gives:

```
function (output_object_list, object_type_label = NULL, excluded_items = c(), 
    as_list = FALSE, list_label = NULL) 
{
    if (is.null(object_type_label)) {
        out <- names(output_object_list)
    }
    else {
        out <- names(output_object_list)[sapply(output_object_list, 
            function(x) any(identical(x$object_type_label, object_type_label)))]
    }
    if (length(out) == 0) {
        return(out)
    }
    if (length(excluded_items) > 0) {
        excluded_indices <- which(out %in% excluded_items)
        if (length(excluded_indices) > 0) {
            out <- out[-excluded_indices]
        }
    }
    if (as_list) {
        lst <- list()
        if (!is.null(list_label)) {
            lst[[list_label]] <- out
        }
        else {
            lst <- as.list(out)
        }
        return(lst)
    }
    else {
        return(out)
    }
}
<bytecode: 0x0000012b97d06300>
<environment: namespace:databook>
```

rdstern commented:
@lilyclements That's interesting.  If I understand correctly, then these aren both working correctly on your machine?
My first theory then is that it might have something to do with the R packages not installed on my machine.  Let me try one thing then next.

lilyclements commented:
Interesting. Yes, they're both working on my machine.

rdstern commented:
@lilyclements just to be clear.  I can import quite some rds files, but not this one:

[eastern_tidy_with_summaries_2.zip](https://github.com/user-attachments/files/18894359/eastern_tidy_with_summaries_2.zip)

which imports with version 0.8.2 but not with this.

And can you check with a csv file.  I could not get any of them to show a preview and hence Ok wasn't enabled.

With the Excel files it gave that nasty message, but Ok was enabled and it imported fine - one sheet or multiple sheets.

lilyclements commented:
This is great - we're on the same page! I will look tomorrow at

1. Importing this RDS file you shared
2. Getting the viewer up for excel files (I didn't even clock it wasn't there, I was focussing too hard on the OK being enabled)
3. Getting CSV files to work again
4. Checking the code for other formats

rdstern commented:
@lilyclements thanks for letting me know.  That's a big relief - it is all so much harder if the problems are only on some machines!

lilyclements commented:
@rdstern after a good few hours of hunting, I think I've found out what the issue is, and I don't like it.

Since our `.rds` files were saved with an old version of `DataBook` (when it was still tied to R-Instat), it is trying to find old method references when we load them in - references to now non-existent code in R-Instat, not code from the new `databook` package. This definitely needs to involve a conversation with David to help iron out and work out the best approach going forward. 

For now, some things need to remain in the environment to avoid this error and help in testing. Only very few bits, and if you do get another error then let me know and I can amend as needed. 

I'll discuss with David how we can sort this out going forward. Perhaps when importing if there's an error we can detect if it's because it's with a RDS file attached to the R-Instat code instead of the databook code, and offer to update and resave it accordingly. I am not sure how this works on the code side, but is certainly something i need to discuss with David regardless on the very least of how we can conceptually handle this.

For the RDS files I was having issues with, it is now working with this temporary solution. This is a great bug to find, but definitely not one I was expecting at all.

Next step: I'll sort out the .csv and .xlsx files.

rdstern commented:
@lilyclements delighted you found that, and that's a good logical reason for that problem.  Interesting that the olde rds files, e.g. Ghana climatic are ok, and the newer ones, e.g. Zambia, are not.

And the good news remains that I have now tried a lot of tother things, and they seem to work.

rdstern commented:
@lilyclements I have just been using your branch, and have yet to update, because of your Friday work.  But, but, but I was importing some Excel data and look at this:

![image](https://github.com/user-attachments/assets/a6beddbd-9909-4845-ba6e-d03785424af9)

It is perfect!  Why is it suddenly working.  I was assuming I would blindly press Ok, because there is no preview.  Now, lo and behold it is there?  Let me continue.  I'm combining testing just now, with some Zimbabwe work!  But this is very odd.

By the way, up above I also have a question on the different R-Instat packages.

rdstern commented:
@lilyclements I don't understand anymore!  Here is your branch importing a new Excel file.  And it looks perfect!  I noticed this yesterday when I needed to upload 4 sheets.  I thought I could read them without having the preview.  Now, lo and behold, here is the preview!

![image](https://github.com/user-attachments/assets/b647374e-3f15-4207-932a-ac88215cd232)

This is before I update the packages.  I wonder which of the 4 you have changed and whether I can update from R-Instat, or need to go to RStudio?

And here is the first csv I have come to:

![image](https://github.com/user-attachments/assets/9eb93c61-527c-430e-beec-e8a214d468fe)

And the rds files that gave the error, now import fine too.  What is going on?

lilyclements commented:
@N-thony happy to archive instead of remove, provided they don't run on upload and save to the environment. I have been removing in the time being to ensure it is _definitely_ using our package, and not the VB code.

To update or fix R code, we will now go through the packages. This is actually simpler in debugging than we've had before now (in my opinion..). Roger and I have agreed we will need to arrange a meeting with everyone to go through this new system - and I'll create some resources too that explains in (resources = word document?). What else would be helpful?

lilyclements commented:
@rdstern I'm glad that's all coming in fine now - I hope that is because of some changes I made on the branch. I spoke briefly with David on the RDS file, and he seems to agree with me on how we can fix it, but we definitely need another conversation on it (an actual meeting!)

ChrisMarsh82 commented:
> @lilyclements just a few concerns I have
> 
> 1. What will happen if we have an issue with one of the R functions that need to be updated or bug fix?
> 2. I see you have removed the code in the `instat_object` and `data_object` files. Won't it be nicer instead to archive the files?

I don't see what you mean by archive? If the code is not used delete it. Git will always allow you to go back to an older version if needed. 

@lilyclements Regarding what we do if there is an issue. Do you have any unit tests on the functions. This could help with future updates as we make sure we are not breaking existing functionality

ChrisMarsh82 commented:
@N-thony Can you update install_packages and InstallPackages and link the PR to this one.

rdstern commented:
@N-thony  and @berylwaswa and @lilyclements I suggest a plan also to help with your justified concerns, Antoine.

I was just writing our cautious plan, but now have an alternative, see below:
a) Lily is away in a meeting all this week.  I suggest I will continue to test her branch and it would be good if you and Beryl could do the same.  Antoine I suggest you are more an installer of her branch and then, initially, a tester here, rather than a code checker.  That can come later.
b) To test, you just go to the branch as any other, but then you need to include the 4 main R-packages first.  I hope this will just use the Tools > Install R Package, and the second button at then top.

There are 4 packages, not yet in the drop-down list - so type the names:
instatCalculations, instatExtras, instatClimatic and databook

Once installed, as shown below, the check goes brown:

![image](https://github.com/user-attachments/assets/3843400a-63e3-4464-a8bc-10d6799fb330)

I hope they install easily for you.  For me some stayed green, because they didn't install properly.  The errors were sometimes shown in Visual Studio.  Otherwise I installed from RStudio.

An alternative is to produce a trial version.  Not for the website, but to make the checking easier to do?  I suggest updating the packages is easier in a "proper" version.  I have the impression that @N-thony can now produce a trial version reasonably quickly.  Updates to the packages can then be added.

This trial version could be useful for another reason.  We have just merged the outfilling dialog, and this is needed next week by @jkmusyoka .  If it is working ok, then he could use this version in Zambia?

lilyclements commented:
> @lilyclements, I'm updating the installation scripts for packages. Do we always need this line?

@N-thony no we don't need this line. We can also remove the line `data_book <- DataBook$new()` because that gets read in in the environment when we call in the `databook` package now

rdstern reviewed (approved):
@lilyclements I am now suggesting we merge this tomorrow and also produce a new version of R-Instat.  This could be 0.8.3, but is mainly designed for testing and also to be used by @jkmusyoka in Zambia next week.  He needs the new outfilling dialog that was merged yesterday.

@N-thony if you are happy to do this, then I suggest the following strategy:

a) Could you check the pull requests I have approved and merge what you can,  They are:
#9454 , #9463, #9458 , #9456 , #9446 , #9452 and #9319 (if @ChrisMarsh82 approves).

b) Then merge this one.  And produce the next version.  

c) Then @jkmusyoka , @berylwaswa and I will test this new version until the end of the week.   The developers continue with their work, but no other pull requests are merged until next week.

How does that sound?
